testinfra/backend/base.py

```diff
@@ -193,7 +194,8 @@ def run(self, command, *args, **kwargs):
 
     def run_local(self, command, *args):
         command = self.quote(command, *args)
-        command = self.encode(command)
+        if os.name != 'nt':
+            command = self.encode(command)
```

Comment:
I'd certainly be willing to believe that there's encoding issues on different platforms.

But the original error seems to be

```
TypeError: argument of type 'int' is not iterable
```

I can not see how a priori how what seems like an "integer has crept in that should be a string" issue relates to string encoding.   So this isn't an obvious merge.

That said, because encoding issues on commandline args for Windows is something I imagine nobody is super conversant in, this change should come with even more explanation than usual in the change, and probably even an inline comment.  Please if you're putting things in for maintainers to look at, add enough context in descriptions for it to be actionable.

Reply:
Hi,

Sorry - probably should have added a bit more context, was mostly just creating a PR for the changes proposed in someone else's issue.

I've added a comment explaining that it's an internal thing inside subprocess to the code. A fuller explanation is:
subprocess.Popen() accepts either "a sequence of program arguments or else a single string" as a command to run.
On POSIX, a string means "the exact command, with no arguments" whereas a byte sequence means "the command and arguments" - I guess this is why testinfra encodes the string in general.
On Windows, a string means "the exact, escaped command, including arguments", and a list means "a command with args that needs to be escaped/joined". 

The fun comes on Windows as the encode() causes the following:
string "echo hello" passed to run_local()
string converted to b"echo hello"
b"echo hello" passed to subprocess.Popen
Popen checks the type, finds it's not a string, and runs list2cmdline()
list2cmdline expects a list, but gets a sequence of bytes, eventually blowing up at https://github.com/python/cpython/blob/v3.7.3/Lib/subprocess.py#L530 as iterating over a sequence of bytes leads to integers.

Reply:
Thanks for the info, TIL about windows process creation functions and command-line encoding types, which I think everyone would be happier never having to think about ever :)

I tried to have a poke and see if anything obvious fell out, but I just ended up a bit more confused per https://github.com/philpep/testinfra/issues/411#issuecomment-521466951

Maybe this is right, but I suspect more that keeping this as "bytes" internally is wrong...

Reply:
Yeah, maybe "bytes" isn't correct internally. I guess some of this may come out with further testing - my team and I may well be doing a bit more work on Windows in the near future, so might end up doing some more stuff around getting the testinfra test suite working, which may reveal more things.